include/revng/ADT/STLExtras.h

```diff
@@ -151,13 +151,41 @@ using MapToValueIteratorType = decltype(mapToValueIterator(std::declval<T>()));
 
 } // namespace revng
 
-template<typename C>
-inline auto skip(unsigned ToSkip, C &&Container)
-  -> llvm::iterator_range<decltype(Container.begin())> {
-  auto Begin = std::begin(Container);
-  while (ToSkip-- > 0)
-    Begin++;
-  return llvm::make_range(Begin, std::end(Container));
+//
+// skip
+//
+
+template<typename T>
+inline auto skip(T &&Range, std::size_t Front = 0, std::size_t Back = 0)
+  -> llvm::iterator_range<decltype(Range.begin())> {
+
+  std::ptrdiff_t TotalSkippedCount = Front + Back;
+  revng_assert(std::distance(Range.begin(), Range.end()) >= TotalSkippedCount);
+
+  auto Begin = std::begin(Range);
+  std::advance(Begin, Front);
+
+  auto End = std::end(Range);
+  std::advance(End, -Back);
+
+  return llvm::make_range(Begin, End);
+}
+
+// Kept for backwards compatibility.
+// TODO: remove.
+template<typename T>
+inline auto skip(unsigned ToSkip, T &&Range) {
+  return skip(Range, ToSkip);
+}
+
+template<typename T> // NOLINTNEXTLINE
+inline auto skip_front(T &&Range, std::size_t SkippedCount = 1) {
```

Comment:
You use `std::advance` with a negative value, which requires a bidirectional iterator.

> Why would we want to limit it to forwards-only?

Right now we're limited to bidirectional iterators, I want to *extend* to forward only iterators.

Reply:
Technically, it works with `forward_iterator`s already: https://godbolt.org/z/8G1vY8Mbd
Because, you know, `std::advance` with negative value is never called in `skip_front`.

BUT, thanks to investigating this, I also figured out that `skip_back` quietly aborts when called with a `std::forward_iterator` which is not very :enterprise:

Reply:
Ok, I looked a bit deeper into this and I'd say I'm satisfied with it:
```cpp
namespace revng::detail {

template<typename IteratorType>
inline auto skipImpl(IteratorType &&From,
                     IteratorType &&To,
                     std::size_t Front = 0,
                     std::size_t Back = 0)
  -> llvm::iterator_range<IteratorType> {

  std::ptrdiff_t TotalSkippedCount = Front + Back;
  /*revng_*/assert(std::distance(From, To) >= TotalSkippedCount);

  std::decay_t<IteratorType> Begin{ From };
  std::advance(Begin, Front);

  std::decay_t<IteratorType> End{ To };
  std::advance(End, -(std::ptrdiff_t)Back);

  return llvm::make_range(std::move(Begin), std::move(End));
}

template<std::bidirectional_iterator T>
inline decltype(auto) skip(T &&From, T &&To, std::size_t Front = 0, std::size_t Back = 0) {
  return skipImpl(std::forward<T>(From), std::forward<T>(To), Front, Back);
}

template<std::input_iterator T> // NOLINTNEXTLINE
inline auto skip_front(T &&From, T &&To, std::size_t SkippedCount = 1) {
  return skipImpl(std::forward<T>(From), std::forward<T>(To), SkippedCount, 0);
}

template<std::bidirectional_iterator T> // NOLINTNEXTLINE
inline auto skip_back(T &&From, T &&To, std::size_t SkippedCount = 1) {
  return skipImpl(std::forward<T>(From), std::forward<T>(To), 0, SkippedCount);
}

} // namespace revng::detail

template<typename T>
inline decltype(auto) skip(T &&Range, std::size_t Front = 0, std::size_t Back = 0) {
  return revng::detail::skip(Range.begin(), Range.end(), Front, Back);
}

template<typename T> // NOLINTNEXTLINE
inline auto skip_front(T &&Range, std::size_t SkippedCount = 1) {
  return revng::detail::skip_front(Range.begin(), Range.end(), SkippedCount);
}

template<typename T> // NOLINTNEXTLINE
inline auto skip_back(T &&Range, std::size_t SkippedCount = 1) {
  return revng::detail::skip_back(Range.begin(), Range.end(), SkippedCount);
}
```
(https://godbolt.org/z/fv5nfo984)

What's your take?

Reply:
Looks good, except for the assertion that won't work with input iterators, I think, since you'll consume the input to compute the assertion and you cannot go back.

Reply:
Good catch!!! 😲

Reply:
I guess this is fine in `STLExtras.h`.

Reply:
I'm closing this thread then.

🤞 Hopefully we get to use `libc++-13`+ soon 🤞